strategies/categorized/src/main/java/org/arquillian/smart/testing/strategies/categorized/AbstractParser.java

```diff
@@ -13,75 +19,154 @@
 
     private final CategorizedConfiguration strategyConfig;
     private final List<String> specifiedCategories;
+    private final List<String> excludedCategories;
+    private final IncludingCategoryFilter includingCategoryFilter;
+    private final ExcludingCategoryFilter excludingCategoryFilter;
 
     AbstractParser(CategorizedConfiguration strategyConfig) {
         this.strategyConfig = strategyConfig;
         specifiedCategories = Arrays.stream(strategyConfig.getCategories())
             .map(this::changeIfNonCaseSensitive)
             .collect(Collectors.toList());
+        excludedCategories = Arrays.stream(strategyConfig.getExcludedCategories())
+            .map(this::changeIfNonCaseSensitive)
+            .collect(Collectors.toList());
+        includingCategoryFilter = new IncludingCategoryFilter();
+        excludingCategoryFilter = new ExcludingCategoryFilter();
     }
 
-    TestSelection getTestSelectionIfMatched(Class<?> clazz){
-        List<String> classCategories = findCategories(clazz.getAnnotations());
-        if (strategyConfig.isMethods() && classCategories.isEmpty()) {
+    TestSelection getTestSelectionIfMatched(Class<?> clazz) {
+        if (strategyConfig.isMethods()) {
             return getSelectionWithMethods(clazz);
-        } else {
-            if (containsCorrectCategoriesMatchingReversed(classCategories)) {
-                return new TestSelection(clazz.getName(), CATEGORIZED);
-            }
+        }
+        return isClassIncluded(clazz) ? new TestSelection(clazz.getName(), CATEGORIZED)
+            : TestSelection.NOT_MATCHED;
+    }
+
+    private boolean isClassIncluded(Class<?> clazz) {
+        return includingCategoryFilter.and(excludingCategoryFilter.negate()).test(findCategories(clazz.getAnnotations()));
+    }
+
+    private TestSelection getSelectionWithMethods(Class<?> clazz) {
+        List<CategorizedMethod> categorizedMethods = getTestMethods(clazz)
+            .stream()
+            .map(method -> new CategorizedMethod(method.getName(), findCategories(method.getAnnotations())))
+            .collect(Collectors.toList());
+        inheritClassLevelCategoriesIfAny(clazz, categorizedMethods);
+        List<String> applicableMethodNames = categorizedMethods.stream()
```

Comment:
here you iterate over the same list of `CategorizedMethod`s twice - in two streams, between these two iterations you feed the `CategorizedMethod`s with the class annotations. 
See my proposal https://github.com/arquillian/smart-testing/pull/311#discussion_r164986429 the idea was that adding the class categories should be part of the `CategorizedMethod` constructor. Another thing to notice is that you retrieve the class categories for every iteration separately https://github.com/arquillian/smart-testing/pull/311/files#diff-b179c0d6894a22602721a54a8efc1d54R70 but this is not necessary as the list of specified categories is always the same. in my proposal is retrieving the class categories once before the loop and feed the  `CategorizedMethod` with it